scvid/module/_probabilisticpca.py

```diff
@@ -0,0 +1,150 @@
+from typing import Optional, Union
+
+import pyro
+import pyro.distributions as dist
+import torch
+from pyro.nn import PyroModule, PyroParam, pyro_method
+from torch.distributions import constraints
+
+_PROBABILISTIC_PCA_PYRO_MODULE_NAME = "probabilistic_pca"
+
+
+class ProbabilisticPCAPyroModule(PyroModule):
+    """
+    Probabilistic PCA implemented in Pyro.
+
+    Args:
+        n_cells: Number of cells.
+        g_genes: Number of genes.
+        k_components: Number of principal components.
+        ppca_flavor: Type of the PPCA model. Has to be one of `marginalize` or `diagonal_normal`
+            or `multivariate_normal`.
+        mean_g: Mean gene expression of the input data.
+        w: Scale of the random initialization of the `W_kg` parameter.
+        s: Initialization value of the `sigma` parameter.
+    """
+
+    def __init__(
+        self,
+        n_cells: int,
+        g_genes: int,
+        k_components: int,
+        ppca_flavor: str,
+        mean_g: Optional[Union[float, int, torch.Tensor]] = None,
+        w: float = 1.0,
+        s: float = 1.0,
+        seed: int = 0,
+    ):
+        super().__init__(_PROBABILISTIC_PCA_PYRO_MODULE_NAME)
+
+        self.n_cells = n_cells
+        self.g_genes = g_genes
+        self.k_components = k_components
+        self.ppca_flavor = ppca_flavor
+
+        if isinstance(mean_g, torch.Tensor) and mean_g.dim():
+            assert mean_g.shape == (
+                g_genes,
+            ), "Expected meang_g to have a shape ({g_genes},) but found {mean_g.shape}."
+        if mean_g is None:
+            # make mean_g a learnable parameter
+            self.mean_g = PyroParam(lambda: torch.zeros(g_genes))
+        else:
+            self.mean_g = mean_g
+
+        rng = torch.Generator()
```

Comment:
Not sure how to handle the `device` at the moment since I don't understand well how it is managed by the PyTorch Lightning. If any issues will come up with the device maybe we can address it then?

Reply:
Sounds good! The [LightningModule](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.core.LightningModule.html#pytorch_lightning.core.LightningModule) knows what device it is on. You can access the reference via `self.device`. A vanilla `nn.Module`, however, does not have `self.device` ...

Reply:
I see. Can you approve this PR if everything looks ok to you?

Reply:
Done! ✅

Reply:
Yay!